reverse_words/README.rst

```diff
@@ -0,0 +1,119 @@
+* ``_transpose`` is a convenience function to re-orient the matrices finally coming out of the data stream
+* ``_is_nan`` is a termination function for training - not data processing
+
+Once the data is read in character-wise (``level="character"``), it
+is cleaned up and converted into mini-batches using ``Batch`` with 
+``iteration_scheme=ConstantScheme(10)``, which means that it works on 10 
+sentences simulataneously.  
+
+Each batch of sentences has ``Padding`` applied
+that pads it out to the same length and simultaneously creates ``features_mask`` 
+and ``targets_mask`` overlays that identify which of the data stream's output cells 
+are filled with valid data.  We know that the sentences in each ``Batch`` are 
+less that 100 characters long, however the ``Padding`` only exapands the 
+underlying data to be the size of the longest entry.
+
+
+Structure of the Model
+--------------------------
+
+``class WordReverser`` is initialized with :
+
+* ``dimension`` which refers to the size of the hidden size of the internal state-to-state data
+* ``alphabet_size`` which allows the initial vector embedding (via ``LookupTable``) of characters into 
+  a ``dimension`` dimensional vector (so that ``dimension`` is doing double-duty as the
+  size of the internal recurrence, and the size of the vector embedding)
+
+The model itself is ``Bidirectional``, with ``SimpleRecurrent`` units.  This means
+that the model will contain two chains of SimpleRecurrent units, going forwards and
+backwards along the sequence of characters in the sentence.
+
+At each time-step on the ``encoder`` level (which corresponds to 
+characters in the input stream) :
+
+* the current character is mapped to an embedding vector
+* the input has ``Fork`` applied, so that it can be fed via weights into 
+  the different input paths of ``Birectectional/SimpleRecurrent`` units 
+  (http://blocks.readthedocs.org/en/latest/api/bricks.html?highlight=fork#blocks.bricks.parallel.Fork)
+
+Once the underlying ``Bidirectional/SimpleRecurrent`` unit states are calculated, 
+a ``SequenceGenerator`` scans across them using an attention mechanism 
+(``SequenceContentAttention``).  This ``SequenceGenerator`` contains another 
+recurrent network (``transition``) that converts the ``encoder`` state into
+an output sequence : 
+
+* the attention mechanism calculates an alignment between the internal ``encoder`` 
+  state and the previous output states
+  (http://blocks.readthedocs.org/en/latest/api/bricks.html?highlight=sequencecontentattention#blocks.bricks.attention.SequenceContentAttention)
+* the final output of the ``SequenceGenerator`` is read via a ``SoftmaxEmitter`` and converted to characters 
+* and it feeds this result back to the ``transition`` recurrent level 
+  via a `LookupFeedback`
+
+
+
+Structure of the Training
+--------------------------
+
+The training algorithm uses ``GradientDescent`` with a maximum 
+step-size (``StepClipping(10.0)``) applied to 
```

Comment:
A clarification: `StepClipping(10.0` is the first step applied which clips the gradient to have the norm of at most `10.0`. It is followed by `Scale(0.01)`, which further scales down the clipped gradient.